kernel/thread.c

```diff
@@ -305,7 +305,60 @@ k_tid_t k_thread_create(struct k_thread *new_thread,
 	}
 	return new_thread;
 }
-#endif
+
+
+#ifdef CONFIG_USERSPACE
+_SYSCALL_HANDLER6(k_thread_create,
+		  new_thread, stack_p, stack_size, entry, p1, more_args)
+{
+	int prio;
+	u32_t ret, options;
+	volatile struct _syscall_10_args *margs =
+		(volatile struct _syscall_10_args *)more_args;
+	k_thread_stack_t stack = (k_thread_stack_t)stack_p;
+	u32_t guard_size;
+
+	_SYSCALL_OBJ_INIT(new_thread, K_OBJ_THREAD);
+	_SYSCALL_MEMORY_READ(margs, sizeof(*margs));
+
+	/* Thread needs to have rw access to the entire stack region including
+	 * any guard areas
+	 */
+	guard_size = (u32_t)K_THREAD_STACK_BUFFER(stack) - (u32_t)stack;
+	_SYSCALL_MEMORY_SUM_WRITE(stack, stack_size, guard_size);
+
+	/* Stash struct arguments in local variables to prevent switcheroo
+	 * attacks
+	 */
+	prio = margs->arg8;
+	options = margs->arg9;
+	compiler_barrier(); /* prio and options should be fixed now */
+
+	/* User threads may only create other user threads and they can't
+	 * be marked as essential
+	 */
+	_SYSCALL_VERIFY(options & K_USER);
+	_SYSCALL_VERIFY(!(options & K_ESSENTIAL));
+
+	/* Check validity of prio argument; must be the same or worse priority
+	 * than the caller
+	 */
+	_SYSCALL_VERIFY(_VALID_PRIO(prio, NULL));
+	_SYSCALL_VERIFY(_is_prio_lower_or_equal(prio, _current->base.prio));
+
+	/* Make sure we haven't run out of thread IDs */
+	_SYSCALL_VERIFY(thread_index < 8 * CONFIG_MAX_THREAD_BYTES);
```

Comment:
Even though thread_index is an atomic variable, `atomic_inc()` will only guarantee the atomicity of the increment operation. This won't prevent a race condition from happening when this value is about to hit its limits (so it passes the check), and another thread is successfully created.